torch/_dynamo/utils.py

```diff
@@ -1715,3 +1716,176 @@ def build_checkpoint_variable(**options):
         activation_checkpoint_op,
         **options,
     )
+
+
+# The following functions are adapted from
+# https://github.com/python/cpython/blob/v3.11.4/Lib/traceback.py
+# in order to output source code corresponding to bytecode in 3.11+.
+# We need our own versions since we want to support multiline expressions.
+def _fix_offset(str, offset):
+    as_utf8 = str.encode("utf-8")
+    return len(as_utf8[:offset].decode("utf-8", errors="replace"))
+
+
+@dataclasses.dataclass
+class _Anchors:
+    left_end_lineno: int
+    left_end_offset: int
+    right_start_lineno: int
+    right_start_offset: int
+
+
+# supports anchors on multiple lines
+def _extract_anchors_from_expr(segment):
+    assert sys.version_info >= (3, 11)
+
+    import ast
+
+    try:
+        # expect an expression, so wrap the segment in
+        # brackets to handle multi-line expressions
+        tree = ast.parse("(\n" + segment + "\n)")
+    except SyntaxError:
+        return None
+
+    if len(tree.body) != 1:
+        return None
+
+    lines = segment.split("\n")
+
+    def normalize(lineno, offset):
+        return _fix_offset(lines[lineno - 1], offset)
+
+    def next_valid_char(lineno, col):
+        while lineno < len(lines) and col >= len(lines[lineno]):
+            col = 0
+            lineno += 1
+        assert lineno < len(lines) and col < len(lines[lineno])
+        return lineno, col
+
+    def increment(lineno, col):
+        col += 1
+        lineno, col = next_valid_char(lineno, col)
+        assert lineno < len(lines) and col < len(lines[lineno])
+        return lineno, col
+
+    def nextline(lineno, col):
+        col = 0
+        lineno += 1
+        lineno, col = next_valid_char(lineno, col)
+        assert lineno < len(lines) and col < len(lines[lineno])
+        return lineno, col
+
+    statement = tree.body[0]
+    if isinstance(statement, ast.Expr):
+        expr = statement.value
+        if isinstance(expr, ast.BinOp):
+            cur_lineno = expr.left.end_lineno - 2
```

Comment:
How come -2 here?

Reply:
-1 because line numbers are 1 indexed, and another -1 because we want the location of the last character of `expr.left1` and `end_lineno` is exclusive.